File: src/main/java/ssg/search/collection/es/ESQuery.java
package ssg.search.collection.es;

import com.google.common.base.Joiner;
import org.apache.commons.lang3.StringUtils;
import ssg.framework.domain.FrontUserInfo;
import ssg.search.parameter.Parameter;

public enum ESQuery {
	SRCH_PREFIX_MALL{
		@Override public String getField(Parameter parameter) {
			return "SRCH_PREFIX";
		}

		@Override public String getValue(Parameter parameter){
			return "SCOM";
		}
	},
	SRCH_PREFIX_VIRTUAL{
		@Override public String getField(Parameter parameter) {
			return "SRCH_PREFIX";
		}

		@Override public String getValue(Parameter parameter){
			return "VIRTUAL";
		}
	},
	SRCH_PREFIX_BSHOP{
		@Override public String getField(Parameter parameter) {
			return "SRCH_PREFIX";
		}

		@Override public String getValue(Parameter parameter){
			return "BSHOP";
		}
	},
	SRCH_PREFIX_GLOBAL{
		@Override public String getField(Parameter parameter) {
			return "SRCH_PREFIX";
		}

		@Override public String getValue(Parameter parameter){
			return "GLOBAL";
		}
	},
	SRCH_PREFIX_DISP_FIX_SITE_NO{
		@Override public String getField(Parameter parameter) {
			return "SRCH_PREFIX";
		}

		@Override public String getValue(Parameter parameter){
			String strSiteNo = StringUtils.defaultIfEmpty(parameter.getSiteNo(), "6005");

			if(strSiteNo.equals("6005")){
				return "SCOM";
			}
			else if(strSiteNo.equals("6004")){
				return "SM";
			}
			else if(strSiteNo.equals("6009")){
				return "SD";
			}
			else if(strSiteNo.equals("6001")){
				return "EM";
			}
			else if(strSiteNo.equals("6002")){
				return "TR";
			}
			else if(strSiteNo.equals("6003")){
				return "BOOTS";
			}
			else if(strSiteNo.equals("6100")){
				return "HOWDY";
			}
			else if(strSiteNo.equals("6200")){
				return "STV";
			}
			else if(strSiteNo.equals("6300")){
				return "SIV";
			}
			return "";
		}
	},
	SRCH_PREFIX_DISP{
		@Override public String getField(Parameter parameter) {
			return "SRCH_PREFIX";
		}

		@Override public String getValue(Parameter parameter){
			String strSiteNo = StringUtils.defaultIfEmpty(parameter.getSiteNo(), "6005");
			String strFilterSiteNo = StringUtils.defaultIfEmpty(parameter.getFilterSiteNo(), "6005");

			if(strSiteNo.equals("6005")){
				String value = "SCOM";

				if(strFilterSiteNo.equals("6004")){
					return value + " AND SCOM-SM";
				}
				else if(strFilterSiteNo.equals("6009")){
					return value + " AND SCOM-SD";
				}
				else if(strFilterSiteNo.equals("6001")){
					return value + " AND SCOM-EM";
				}
				else if(strFilterSiteNo.equals("6002")){
					return value + " AND SCOM-TR";
				}
				else if(strFilterSiteNo.equals("6003")){
					return value + " AND SCOM-BT";
				}
				else if(strFilterSiteNo.equals("6100")){
					return value + " AND SCOM-HD";
				}
				else if(strFilterSiteNo.equals("6200")){
					return value + " AND SCOM-TV";
				}
				else if(strFilterSiteNo.equals("6300")){
					return value + " AND SCOM-SI";
				}

				return value;
			}
			else if(strSiteNo.equals("6004")){
				return "SM";
			}
			else if(strSiteNo.equals("6009")){
				return "SD";
			}
			else if(strSiteNo.equals("6001")){
				return "EM";
			}
			else if(strSiteNo.equals("6002")){
				return "TR";
			}
			else if(strSiteNo.equals("6003")){
				return "BOOTS";
			}
			else if(strSiteNo.equals("6100")){
				return "HOWDY";
			}
			else if(strSiteNo.equals("6200")){
				return "STV";
			}
			else if(strSiteNo.equals("6300")){
				return "SIV";
			}
			return "";
		}
	},
	DISP_CTG_LST{
		@Override public String getField(Parameter parameter) {
			return "DISP_CTG_LST";
		}

		@Override public String getValue(Parameter parameter) {
			if(StringUtils.isNotEmpty(parameter.getDispCtgId())){
				return parameter.getDispCtgId();
			}else if(StringUtils.isNotEmpty(parameter.getCtgId())){
				return parameter.getCtgId();
			}

			return "";
		}
	},
	DISP_CTG_SUB_LST{
		@Override public String getField(Parameter parameter) {
			return "DISP_CTG_LST";
		}

		@Override public String getValue(Parameter parameter) {
			Joiner orJoiner = Joiner.on(" OR ").skipNulls();

			if(StringUtils.isNotEmpty(parameter.getDispCtgSubIds())){
				return orJoiner.join(parameter.getDispCtgSubIds().split("\\|"));
			}

			return "";
		}
	},
	BRAND_ID{
		@Override public String getField(Parameter parameter) {
			return "BRAND_ID";
		}

		@Override public String getValue(Parameter parameter) {
			if(StringUtils.isEmpty(parameter.getBrand()) && StringUtils.isEmpty(parameter.getBrandId())){
				return "";
			}
			StringBuilder sb = new StringBuilder();
			String[] brandList;
			if(StringUtils.isNotEmpty(parameter.getBrand())){
				brandList = parameter.getBrand().split("\\|");
			}else{
				brandList = parameter.getBrandId().split("\\|");
			}
			if(brandList.length > 0){
				for(int i = 0; i < brandList.length; i++){
					if(StringUtils.isNotEmpty(brandList[i])){
						if(i != 0 && brandList.length > 1){
							sb.append(" OR ");
						}
						sb.append(brandList[i]);
					}
				}	
			}
			return sb.toString();
		}
	},
	SALESTR_LST{
		@Override public String getField(Parameter parameter) {
			return "SALESTR_LST";
		}

		@Override public String getValue(Parameter parameter) {
			FrontUserInfo userInfo = parameter.getUserInfo();
			String emSalestrNo = StringUtils.defaultIfEmpty(userInfo.getEmSaleStrNo(), "");
			String emRsvtShppPsblYn = StringUtils.defaultIfEmpty(userInfo.getEmRsvtShppPsblYn(),"N");
			String trSalestrNo = StringUtils.defaultIfEmpty(userInfo.getTrSaleStrNo(), "");
			String trRsvtShppPsblYn = StringUtils.defaultIfEmpty(userInfo.getTrRsvtShppPsblYn(),"N");
			String bnSalestrNo = StringUtils.defaultIfEmpty(userInfo.getBnSaleStrNo(), "");
			String bnRsvtShppPsblYn = StringUtils.defaultIfEmpty(userInfo.getBnRsvtShppPsblYn(),"N");
			String strSiteNo = StringUtils.defaultIfEmpty(parameter.getSiteNo(), "6005");
			String deptSalestrNo = StringUtils.defaultIfEmpty(parameter.getSalestrNo(), "0001");
			String hwSalestrNo = StringUtils.defaultIfEmpty(userInfo.getHwSaleStrNo(), "");
			String hwRsvtShppPsblYn = StringUtils.defaultIfEmpty(userInfo.getHwRsvtShppPsblYn(), "N");
			String emDualzSalestrNo = StringUtils.defaultIfEmpty(userInfo.getEmDualSaleStrNo(), "0000");
			String pickuSalestr =  StringUtils.defaultIfEmpty(parameter.getPickuSalestr(), "");

			//매직픽업 점포만 있을경우, picku점포코드 대체(매직픽업 없는 점포의 경우 shpp에서 걸러짐)
			if(!pickuSalestr.equals("") && StringUtils.isEmpty(parameter.getSalestrNo()) )deptSalestrNo = pickuSalestr;

			if(strSiteNo.equals("6005")){
				return "6005 OR " + deptSalestrNo + " OR " + emSalestrNo + emRsvtShppPsblYn + " OR " + trSalestrNo + trRsvtShppPsblYn + " OR " + bnSalestrNo + bnRsvtShppPsblYn + " OR " + hwSalestrNo + hwRsvtShppPsblYn + " OR " +  emSalestrNo + emDualzSalestrNo;
			}
			else if(strSiteNo.equals("6004")){
				return "6005 OR " + deptSalestrNo;
			}
			else if(strSiteNo.equals("6009")){
				return deptSalestrNo;
			}
			else if(strSiteNo.equals("6001") || strSiteNo.equals("6002")){
				return "6005 OR " + emSalestrNo + emRsvtShppPsblYn + " OR " + trSalestrNo + trRsvtShppPsblYn + " OR " + bnSalestrNo + bnRsvtShppPsblYn + " OR " +  emSalestrNo + emDualzSalestrNo;
			}
			else if(strSiteNo.equals("6100")){
				return hwSalestrNo + hwRsvtShppPsblYn;
			}
			else if(strSiteNo.equals("6200")) {
				return "6005";
			}
			else if(strSiteNo.equals("6003")){
				return "6005" +  " OR " + bnSalestrNo + bnRsvtShppPsblYn;
			}
			else if(strSiteNo.equals("6300")){
				return "6005";
			}
			return "";
		}
	},
	SALESTR_LST_MALL{
		@Override public String getField(Parameter parameter) {
			return "SALESTR_LST";
		}

		@Override public String getValue(Parameter parameter) {
			FrontUserInfo userInfo = parameter.getUserInfo();
			String emSalestrNo = StringUtils.defaultIfEmpty(userInfo.getEmSaleStrNo(), "");
			String emRsvtShppPsblYn = StringUtils.defaultIfEmpty(userInfo.getEmRsvtShppPsblYn(),"");
			String trSalestrNo = StringUtils.defaultIfEmpty(userInfo.getTrSaleStrNo(), "");
			String trRsvtShppPsblYn = StringUtils.defaultIfEmpty(userInfo.getTrRsvtShppPsblYn(),"");
			String bnSalestrNo = StringUtils.defaultIfEmpty(userInfo.getBnSaleStrNo(), "");
			String bnRsvtShppPsblYn = StringUtils.defaultIfEmpty(userInfo.getBnRsvtShppPsblYn(),"N");
			String hwSalestrNo = StringUtils.defaultIfEmpty(userInfo.getHwSaleStrNo(), "");
			String hwRsvtShppPsblYn = StringUtils.defaultIfEmpty(userInfo.getHwRsvtShppPsblYn(), "N");
			String emDualzSalestrNo = StringUtils.defaultIfEmpty(userInfo.getEmDualSaleStrNo(), "0000");
			String deptSalestrNo = "0001";

			return "6005 OR " + deptSalestrNo + " OR " + emSalestrNo + emRsvtShppPsblYn + " OR " + trSalestrNo + trRsvtShppPsblYn + " OR " + bnSalestrNo + bnRsvtShppPsblYn + " OR " + hwSalestrNo + hwRsvtShppPsblYn + " OR " +  emSalestrNo + emDualzSalestrNo;
		}
	},
	SIZE{
		@Override public String getField(Parameter parameter) {
			return "SIZE_LST";
		}

		@Override public String getValue(Parameter parameter) {
			if(StringUtils.isEmpty(parameter.getSize())){
				return "";	
			}
			StringBuilder sb = new StringBuilder();
			String[] sizeList = null;
			sizeList = parameter.getSize().split("\\|");
			if(sizeList.length > 0){
				for(int i = 0; i < sizeList.length; i++){
					if(StringUtils.isNotEmpty(sizeList[i])){
						if(i != 0 && sizeList.length > 1){
							sb.append(" OR ");
						}
						sb.append(sizeList[i]);
					}
				}	
			}
			return sb.toString();
		}
	},
	COLOR{
		@Override public String getField(Parameter parameter) {
			return "COLOR_LST";
		}

		@Override public String getValue(Parameter parameter) {
			if(StringUtils.isEmpty(parameter.getColor())){
				return "";	
			}
			StringBuilder sb = new StringBuilder();
			String[] colorList = null;
			colorList = parameter.getColor().split("\\|");
			if(colorList.length > 0){
				for(int i = 0; i < colorList.length; i++){
					if(StringUtils.isNotEmpty(colorList[i])){
						if(i != 0 && colorList.length > 1){
							sb.append(" OR ");
						}
						sb.append(colorList[i]);
					}
				}	
			}
			return sb.toString();
		}
	},
	DEVICE_CD{
		@Override public String getField(Parameter parameter) {
			return "DISP_APL_RNG_TYPE_CD";
		}

		@Override public String getValue(Parameter parameter) {
			if(StringUtils.isNotEmpty(parameter.getAplTgtMediaCd())){
				return parameter.getAplTgtMediaCd();
			}
			return "";
		}
	},
	MBR_CO_TYPE{
		@Override public String getField(Parameter parameter) {
			FrontUserInfo userInfo = parameter.getUserInfo();
			String type = StringUtils.defaultIfEmpty(userInfo.getMbrType(), "");
            if(type.equals("B2C")){
                return "APL_TGT_VEN_LST";
            }else if(type.equals("B2E")){
                return "APL_B2E_MBRCO_LST";
            }
			return "APL_TGT_VEN_LST";
		}

		@Override public String getValue(Parameter parameter) {
			FrontUserInfo userInfo = parameter.getUserInfo();
			String type = StringUtils.defaultIfEmpty(userInfo.getMbrType(), "");
			String value = "";
			if(type.equals("B2E")){
				value += "B2E OR ";
			}

			if(StringUtils.isNotEmpty(userInfo.getMbrCoId())){
                value += userInfo.getMbrCoId();
            }
			return value;
		}
	},
	SPL_VEN_ID{
		@Override public String getField(Parameter parameter) {
			return "SPL_VEN_ID";
		}

		@Override public String getValue(Parameter parameter) {
			if(!StringUtils.isBlank(parameter.getSplVenId())){
				return parameter.getSplVenId();
			}
			return "";
		}
	},
	LRNK_SPL_VEN_ID{
		@Override public String getField(Parameter parameter) {
			return "LRNK_SPL_VEN_ID";
		}

		@Override public String getValue(Parameter parameter) {
			if(StringUtils.isNotEmpty(parameter.getSplVenId())){
				return parameter.getSplVenId();
			}
			return "";
		}
	},
	SCOM_EXPSR_YN{
		@Override public String getField(Parameter parameter) {
			return "SCOM_EXPSR_YN";
		}

		@Override public String getValue(Parameter parameter) {
			String strSiteNo = StringUtils.defaultIfEmpty(parameter.getSiteNo(), "6005");
			String strFilterSiteNo = StringUtils.defaultIfEmpty(parameter.getFilterSiteNo(), "");
			if(strSiteNo.equals("6005") && ( strFilterSiteNo.equals("") || strFilterSiteNo.equals("6005")) ){
				return "Y";
			}
			return "";
		}
	},
	PRC_FILTER{
		@Override public String getField(Parameter parameter) {
			return "SELLPRC";
		}

		@Override public String getValue(Parameter parameter) {
			String strMinPrc = StringUtils.defaultIfEmpty(parameter.getMinPrc(), "*");
			String strMaxPrc = StringUtils.defaultIfEmpty(parameter.getMaxPrc(), "*");

			if (StringUtils.equals(strMinPrc, "*") && StringUtils.equals(strMaxPrc, "*")) {
				return "";
			} else {
				return "[ " + strMinPrc + " TO " + strMaxPrc + " ]";
			}
		}
	},
	GRP_ADDR_ID{
		@Override public String getField(Parameter parameter) {
			return "GRP_ADDR_ID";
		}

		@Override public String getValue(Parameter parameter) {
			if(StringUtils.isNotEmpty(parameter.getGrpAddrId())){
				return parameter.getGrpAddrId();
			}
			return "";
		}
	},
	SHPPCST_ID{
		@Override public String getField(Parameter parameter) {
			return "SHPPCST_ID";
		}

		@Override public String getValue(Parameter parameter) {
			if(StringUtils.isNotEmpty(parameter.getShppcstId())){
				return parameter.getShppcstId();
			}
			return "";
		}
	},
	ITEM_SITE_NO{
		@Override public String getField(Parameter parameter) {
			return "SITE_NO";
		}

		@Override public String getValue(Parameter parameter) {
			if(StringUtils.equals(parameter.getTarget(), "es_bundle") && !StringUtils.equals(parameter.getSiteNo(), "6005") && StringUtils.isNotEmpty(parameter.getItemSiteNo())){
				return parameter.getItemSiteNo();
			}
			return "";
		}
	},
	BSHOP_ID{
		@Override public String getField(Parameter parameter) {
			return "BSHOPID_LST";
		}

		@Override public String getValue(Parameter parameter) {
			if(StringUtils.isNotEmpty(parameter.getBshopId())){
				return parameter.getBshopId();
			}
			return "";
		}
	};

	public abstract String getValue(Parameter parameter);
	public abstract String getField(Parameter parameter);
}

File: src/main/kotlin/ssg/front/search/api/collection/es/EsQuery.kt
package ssg.front.search.api.collection.es

import com.google.common.base.Joiner
import ssg.front.search.api.dto.FrontUserInfo
import ssg.front.search.api.dto.Parameter

enum class EsQuery {
    SRCH_PREFIX_MALL {
        override fun getField(parameter: Parameter): String {
            return "SRCH_PREFIX"
        }

        override fun getValue(parameter: Parameter): String {
            return "SCOM"
        }
    },
    SRCH_PREFIX_VIRTUAL {
        override fun getField(parameter: Parameter): String {
            return "SRCH_PREFIX"
        }

        override fun getValue(parameter: Parameter): String {
            return "VIRTUAL"
        }
    },
    SRCH_PREFIX_BSHOP {
        override fun getField(parameter: Parameter): String {
            return "SRCH_PREFIX"
        }

        override fun getValue(parameter: Parameter): String {
            return "BSHOP"
        }
    },
    SRCH_PREFIX_GLOBAL {
        override fun getField(parameter: Parameter): String {
            return "SRCH_PREFIX"
        }

        override fun getValue(parameter: Parameter): String {
            return "GLOBAL"
        }
    },
    SRCH_PREFIX_DISP_FIX_SITE_NO {
        override fun getField(parameter: Parameter): String {
            return "SRCH_PREFIX"
        }

        override fun getValue(parameter: Parameter): String {
            val strSiteNo = parameter.siteNo

            when (strSiteNo) {
                "6005" -> return "SCOM"
                "6004" -> return "SM"
                "6009" -> return "SD"
                "6001" -> return "EM"
                "6002" -> return "TR"
                "6003" -> return "BOOTS"
                "6100" -> return "HOWDY"
                "6200" -> return "STV"
                "6300" -> return "SIV"
                else -> return ""
            }
        }
    },
    SRCH_PREFIX_DISP {
        override fun getField(parameter: Parameter): String {
            return "SRCH_PREFIX"
        }

        override fun getValue(parameter: Parameter): String {
            val strSiteNo = parameter.siteNo
            val strFilterSiteNo = parameter.filterSiteNo ?: "6005"

            when (strSiteNo) {
                "6005" -> {
                    val value = "SCOM"

                    if (strFilterSiteNo == "6004") {
                        return "$value AND SCOM-SM"
                    } else if (strFilterSiteNo == "6009") {
                        return "$value AND SCOM-SD"
                    } else if (strFilterSiteNo == "6001") {
                        return "$value AND SCOM-EM"
                    } else if (strFilterSiteNo == "6002") {
                        return "$value AND SCOM-TR"
                    } else if (strFilterSiteNo == "6003") {
                        return "$value AND SCOM-BT"
                    } else if (strFilterSiteNo == "6100") {
                        return "$value AND SCOM-HD"
                    } else if (strFilterSiteNo == "6200") {
                        return "$value AND SCOM-TV"
                    } else if (strFilterSiteNo == "6300") {
                        return "$value AND SCOM-SI"
                    }

                    return value
                }
                "6004" -> return "SM"
                "6009" -> return "SD"
                "6001" -> return "EM"
                "6002" -> return "TR"
                "6003" -> return "BOOTS"
                "6100" -> return "HOWDY"
                "6200" -> return "STV"
                "6300" -> return "SIV"
                else -> return ""
            }
        }
    },
    DISP_CTG_LST {
        override fun getField(parameter: Parameter): String {
            return "DISP_CTG_LST"
        }

        override fun getValue(parameter: Parameter): String {
            if (!parameter.dispCtgId.isNullOrEmpty()) {
                return parameter.dispCtgId!!
            } else if (!parameter.ctgId.isNullOrEmpty()) {
                return parameter.ctgId!!
            }

            return ""
        }
    },
    DISP_CTG_SUB_LST {
        override fun getField(parameter: Parameter): String {
            return "DISP_CTG_LST"
        }

        override fun getValue(parameter: Parameter): String {
            val orJoiner = Joiner.on(" OR ").skipNulls()

            return if (!parameter.dispCtgSubIds.isNullOrEmpty()) {
                orJoiner.join(parameter.dispCtgSubIds!!.split("\\|".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray())
            } else ""

        }
    },
    BRAND_ID {
        override fun getField(parameter: Parameter): String {
            return "BRAND_ID"
        }

        override fun getValue(parameter: Parameter): String {
            if (parameter.brand.isNullOrEmpty() && parameter.brandId.isNullOrEmpty()) {
                return ""
            }
            val sb = StringBuilder()
            val brandList: Array<String> = if (!parameter.brand.isNullOrEmpty()) {
                parameter.brand!!.split("\\|".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            } else {
                parameter.brandId!!.split("\\|".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            }
            if (brandList.isNotEmpty()) {
                for (i in brandList.indices) {
                    if (!brandList[i].isNullOrEmpty()) {
                        if (i != 0 && brandList.size > 1) {
                            sb.append(" OR ")
                        }
                        sb.append(brandList[i])
                    }
                }
            }
            return sb.toString()
        }
    },
    SALESTR_LST {
        override fun getField(parameter: Parameter): String {
            return "SALESTR_LST"
        }

        override fun getValue(parameter: Parameter): String {
            val userInfo = parameter.userInfo
            val emSalestrNo = userInfo.emSaleStrNo
            val emRsvtShppPsblYn = userInfo.emRsvtShppPsblYn
            val trSalestrNo = userInfo.trSaleStrNo
            val trRsvtShppPsblYn = userInfo.trRsvtShppPsblYn
            val bnSalestrNo = userInfo.bnSaleStrNo
            val bnRsvtShppPsblYn = userInfo.bnRsvtShppPsblYn
            val strSiteNo = parameter.siteNo
            var deptSalestrNo = parameter.salestrNo ?: "0001"
            val hwSalestrNo = userInfo.hwSaleStrNo
            val hwRsvtShppPsblYn = userInfo.hwRsvtShppPsblYn
            val emDualzSalestrNo = userInfo.emDualSaleStrNo ?: "0000"
            val pickuSalestr = parameter.pickuSalestr ?: ""

            //매직픽업 점포만 있을경우, picku점포코드 대체(매직픽업 없는 점포의 경우 shpp에서 걸러짐)
            if (pickuSalestr != "" && parameter.salestrNo.isNullOrEmpty()) deptSalestrNo = pickuSalestr

            if (strSiteNo == "6005") {
                return "6005 OR $deptSalestrNo OR $emSalestrNo$emRsvtShppPsblYn OR $trSalestrNo$trRsvtShppPsblYn OR $bnSalestrNo$bnRsvtShppPsblYn OR $hwSalestrNo$hwRsvtShppPsblYn OR $emSalestrNo$emDualzSalestrNo"
            } else if (strSiteNo == "6004") {
                return "6005 OR $deptSalestrNo"
            } else if (strSiteNo == "6009") {
                return deptSalestrNo
            } else if (strSiteNo == "6001" || strSiteNo == "6002") {
                return "6005 OR $emSalestrNo$emRsvtShppPsblYn OR $trSalestrNo$trRsvtShppPsblYn OR $bnSalestrNo$bnRsvtShppPsblYn OR $emSalestrNo$emDualzSalestrNo"
            } else if (strSiteNo == "6100") {
                return hwSalestrNo + hwRsvtShppPsblYn
            } else if (strSiteNo == "6200") {
                return "6005"
            } else if (strSiteNo == "6003") {
                return "6005 OR $bnSalestrNo$bnRsvtShppPsblYn"
            } else if (strSiteNo == "6300") {
                return "6005"
            }
            return ""
        }
    },
    SALESTR_LST_MALL {
        override fun getField(parameter: Parameter): String {
            return "SALESTR_LST"
        }

        override fun getValue(parameter: Parameter): String {
            var userInfo = FrontUserInfo()
            val emSalestrNo = userInfo.emSaleStrNo ?: ""
            val emRsvtShppPsblYn = userInfo.emRsvtShppPsblYn ?: ""
            val trSalestrNo = userInfo.trSaleStrNo ?: ""
            val trRsvtShppPsblYn = userInfo.trRsvtShppPsblYn ?: ""
            val bnSalestrNo = userInfo.bnSaleStrNo ?: ""
            val bnRsvtShppPsblYn = userInfo.bnRsvtShppPsblYn ?: "N"
            val hwSalestrNo = userInfo.hwSaleStrNo ?: ""
            val hwRsvtShppPsblYn = userInfo.hwRsvtShppPsblYn ?: "N"
            val emDualzSalestrNo = userInfo.emDualSaleStrNo ?: "0000"
            var deptSalestrNo = "0001"

            return "6005 OR $deptSalestrNo OR $emSalestrNo$emRsvtShppPsblYn OR $trSalestrNo$trRsvtShppPsblYn OR $bnSalestrNo$bnRsvtShppPsblYn OR $hwSalestrNo$hwRsvtShppPsblYn OR $emSalestrNo$emDualzSalestrNo"
        }
    },
    SIZE {
        override fun getField(parameter: Parameter): String {
            return "SIZE_LST"
        }

        override fun getValue(parameter: Parameter): String {
            if (parameter.size.isNullOrEmpty()) {
                return ""
            }
            val sb = StringBuilder()
            var sizeList = parameter.size!!.split("\\|".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            if (sizeList.isNotEmpty()) {
                for (i in sizeList.indices) {
                    if (!sizeList[i].isNullOrEmpty()) {
                        if (i != 0 && sizeList.size > 1) {
                            sb.append(" OR ")
                        }
                        sb.append(sizeList[i])
                    }
                }
            }
            return sb.toString()
        }
    },
    COLOR {
        override fun getField(parameter: Parameter): String {
            return "COLOR_LST"
        }

        override fun getValue(parameter: Parameter): String {
            if (parameter.color.isNullOrEmpty()) {
                return ""
            }
            val sb = StringBuilder()
            var colorList = parameter.color!!.split("\\|".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            if (colorList.isNotEmpty()) {
                for (i in colorList.indices) {
                    if (!colorList[i].isNullOrEmpty()) {
                        if (i != 0 && colorList.size > 1) {
                            sb.append(" OR ")
                        }
                        sb.append(colorList[i])
                    }
                }
            }
            return sb.toString()
        }
    },
    DEVICE_CD {
        override fun getField(parameter: Parameter): String {
            return "DISP_APL_RNG_TYPE_CD"
        }

        override fun getValue(parameter: Parameter): String {
            return if (!parameter.aplTgtMediaCd.isNullOrEmpty()) {
                parameter.aplTgtMediaCd!!
            } else ""
        }
    },
    MBR_CO_TYPE {
        override fun getField(parameter: Parameter): String {
            var userInfo = FrontUserInfo()
            val type = userInfo.mbrType ?: ""
            if (type == "B2C") {
                return "APL_TGT_VEN_LST"
            } else if (type == "B2E") {
                return "APL_B2E_MBRCO_LST"
            }
            return "APL_TGT_VEN_LST"
        }

        override fun getValue(parameter: Parameter): String {
            var userInfo = FrontUserInfo()
            val type = userInfo.mbrType ?: ""
            var value = ""
            if (type == "B2E") {
                value += "B2E OR "
            }

            if (!userInfo.mbrCoId.isNullOrEmpty()) {
                value += userInfo.mbrCoId
            }
            return value
        }
    },
    SPL_VEN_ID {
        override fun getField(parameter: Parameter): String {
            return "SPL_VEN_ID"
        }

        override fun getValue(parameter: Parameter): String {
            return if (!parameter.splVenId.isNullOrEmpty()) {
                parameter.splVenId!!
            } else ""
        }
    },
    LRNK_SPL_VEN_ID {
        override fun getField(parameter: Parameter): String {
            return "LRNK_SPL_VEN_ID"
        }

        override fun getValue(parameter: Parameter): String {
            return if (!parameter.splVenId.isNullOrEmpty()) {
                parameter.splVenId!!
            } else ""
        }
    },
    SCOM_EXPSR_YN {
        override fun getField(parameter: Parameter): String {
            return "SCOM_EXPSR_YN"
        }

        override fun getValue(parameter: Parameter): String {
            val strSiteNo = parameter.siteNo
            val strFilterSiteNo = parameter.filterSiteNo ?: ""
            return if (strSiteNo == "6005" && (strFilterSiteNo == "" || strFilterSiteNo == "6005")) {
                "Y"
            } else ""
        }
    },
    PRC_FILTER {
        override fun getField(parameter: Parameter): String {
            return "SELLPRC"
        }

        override fun getValue(parameter: Parameter): String {
            val strMinPrc = parameter.minPrc ?: "*"
            val strMaxPrc = parameter.maxPrc ?: "*"

            return if (strMinPrc == "*" && strMaxPrc == "*") {
                ""
            } else {
                "[ $strMinPrc TO $strMaxPrc ]"
            }
        }
    },
    GRP_ADDR_ID {
        override fun getField(parameter: Parameter): String {
            return "GRP_ADDR_ID"
        }

        override fun getValue(parameter: Parameter): String {
            return if (!parameter.grpAddrId.isNullOrEmpty()) {
                parameter.grpAddrId!!
            } else ""
        }
    },
    SHPPCST_ID {
        override fun getField(parameter: Parameter): String {
            return "SHPPCST_ID"
        }

        override fun getValue(parameter: Parameter): String {
            return if (!parameter.shppcstId.isNullOrEmpty()) {
                parameter.shppcstId!!
            } else ""
        }
    },
    ITEM_SITE_NO {
        override fun getField(parameter: Parameter): String {
            return "SITE_NO"
        }

        override fun getValue(parameter: Parameter): String {
            return if (parameter.target == "es_bundle" && parameter.siteNo != "6005" && !parameter.itemSiteNo.isNullOrEmpty()) {
                parameter.itemSiteNo!!
            } else ""
        }
    },
    BSHOP_ID {
        override fun getField(parameter: Parameter): String {
            return "BSHOPID_LST"
        }

        override fun getValue(parameter: Parameter): String {
            return if (!parameter.bshopId.isNullOrEmpty()) {
                parameter.bshopId!!
            } else ""
        }
    };

    abstract fun getValue(parameter: Parameter): String
    abstract fun getField(parameter: Parameter): String
}